### PR TITLE
chore(config): initial integration of configuration schema for sources

### PR DESCRIPTION
**Pull request by tobz**
(part of the work required for #12223)

This PR is the initial implementation of the configuration schema framework, specifically for sources.

It does not change how we load configurations, or anything about our use of `inventory`/`typetag`/`SourceOuter`/etc. That work will be the final step after all components have been minimally instrumented/converted to support having their configuration schema generated.

## Reviewer Notes

In general, the a small portion of this PR is adding the `#[configurable_component]` attribute macro to the configuration types for sources, including any types that they transitively depend on. Additionally, due to a requirement by the `Configurable` derive, all derived types must have public documentation (sometimes specified on fields directly, sometimes derived from the field type itself) which represents the vast majority of actual changed lines in this PR.

I've strived to port over the existing Cue documentation in a faithful manner, but there are a number of small divergences:
- making the documentation of common fields (such as `host_key` or `tls.*`, etc) consistency across sources
- fixing spelling/grammatical issues, changing from passive to active voice
- some of the Cue-specific links have stayed as their Cue syntax (because there was no better alternative at the moment) but I've switched to normal Markdown links where possible

### CODE REVIEW

File: src/sources/socket/tcp.rs [hunk 1]
```diff
@@ -14,21 +14,56 @@ use crate::{
...
+
+    /// The timeout, in seconds, before a connection is forcefully closed during shutdown.
     #[serde(default = "default_shutdown_timeout_secs")]
     shutdown_timeout_secs: u64,
+
+    /// Overrides the name of the log field used to add the peer host to each event.
+    ///
+    /// The value will be the peer host's address, including the port i.e. `1.2.3.4:9000`.
+    ///
+    /// By default, the [global `host_key` option](\(urls.vector_configuration)/global-options#log_schema.host_key) is used.
```
Comment (jszwedko) [inline src/sources/socket/tcp.rs:40]: Just noting that I think we'll want to replace these references to urls defined in the CUE docs.

**Review by jszwedko — APPROVED** [review]
This is coming along nicely!

For other reviewers, you can run `cargo test vector_config -- --nocapture` to see the generated schema.

I skimmed over most of the doc comments and focused more on the generated schema which looks good to me. I expect it'll drift if we modify the cue docs, but is a big step in the right direction. Hopefully we can drop the CUE definitions sooner than later (maybe we could generate the CUE definition from the JSON schema until the docs switch over?).

File: lib/vector-config-macros/src/lib.rs [hunk 2]
```diff
@@ -1,18 +1,44 @@
...
+/// ```
+///
+/// This adds special metadata to the generated schema for that type indicating that it represents the configuration of
+/// a component of the specified type.
 #[proc_macro_attribute]
 pub fn configurable_component(args: TokenStream, item: TokenStream) -> TokenStream {
     configurable_component::configurable_component_impl(args, item)
 }
 
+/// A helpful lil derive.
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply lib/vector-config-macros/src/lib.rs:41]: I mean I'd say it's a lil helpful 🥁

**Review by lukesteensen — APPROVED** [review]
Nice! This is looking really neat.

File: lib/codecs/src/decoding/framing/character_delimited.rs [hunk 3]
```diff
@@ -28,7 +29,8 @@ impl CharacterDelimitedDecoderConfig {
 }
 
 /// Options for building a `CharacterDelimitedDecoder`.
-#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
+#[configurable_component]
```
[the remark this answers is not in the document]
Comment (tobz) [reply lib/codecs/src/decoding/framing/character_delimited.rs:32]: Effectively, yes.

Specifically, deriving `Configurable` requires all fields/variants on the container (struct, enum) to also implement `Configurable`.

File: lib/vector-common/src/datetime.rs [hunk 4]
```diff
@@ -82,3 +87,19 @@ pub mod ser_de {
         }
     }
 }
+
+impl<'de> Configurable<'de> for TimeZone {
+    fn referencable_name() -> Option<&'static str> {
+        Some("vector_common::TimeZone")
+    }
+
+    fn description() -> Option<&'static str> {
```
[the remark this answers is not in the document]
Comment (tobz) [reply lib/vector-common/src/datetime.rs:96]: It doesn't make sense for a lot of integral/scalar types like `String` or `u64` or whatever... those types are already represented in JSON Schema as `string` or `number` and we never define custom schema definitions for them.

That's why it's optional.  Additionally, it allows for things like newtype wrappers where we can omit needing a description on the newtype wrapper itself and delegate it to the inner type, etc.

We enforce descriptions on derived implementations of `Configurable` for structs and enums themselves, but fields can be more compile-time fluid, delegating to the field type, etc, as mentioned above, and those checks happen at runtime, so the trait has to be flexible enough to accommodate that.

File: src/sources/socket/tcp.rs (hunk 1 above)
Comment (tobz) [reply src/sources/socket/tcp.rs:40]: Yes, agreed. I had that thought as I was doing full-on Markdown links that I might as well just link to the existing Cue docs via Markdown.

I'll get that fixed up.

File: lib/vector-config-macros/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tobz) [reply lib/vector-config-macros/src/lib.rs:41]: Haha, woops! Will fix.

File: lib/vector-common/src/datetime.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply lib/vector-common/src/datetime.rs:96]: Ah, gotcha, yeah, I see. That makes sense for scalar types.

**Pull request merged**